src/Carnac.Logic/MessageController.cs

```diff
@@ -26,9 +28,13 @@ public KeysController(ObservableCollection<Message> keys, IMessageProvider messa
 
         public void Start()
         {
-            var messageStream = messageProvider.GetMessageStream(keyProvider.GetKeyStream()).Publish().RefCount();
+            var keyStream = keyProvider.GetKeyStream();
+            var messageStream = messageProvider.GetMessageStream(keyStream).Publish();
 
-            var addMessageStream = messageStream.Select(m => Tuple.Create(m, ActionType.Add));
+            var addMessageSubscription = messageStream
+                .ObserveOn(concurrencyService.MainThreadScheduler)
+                .SubscribeOn(concurrencyService.MainThreadScheduler)
```

Comment:
hrrmm, i need to move this before the publish don't I.

It is the messageProvider's subscription which needs to be on the UI thread.